Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,625 @@
+          μ (x ^ ((m1 : ℕ) * (n / (m1 : ℕ))) * x ^ (n % m1)) ^ (1 / (n : ℝ)) := by rw [pow_add]
+      _ ≤ (μ (x ^ ((m1 : ℕ) * (n / (m1 : ℕ)))) * μ (x ^ (n % m1))) ^ (1 / (n : ℝ)) :=
+        (rpow_le_rpow (apply_nonneg μ _) (map_mul_le_mul μ _ _) (one_div_cast_nonneg _))
+      _ = μ (x ^ ((m1 : ℕ) * (n / (m1 : ℕ)))) ^ (1 / (n : ℝ)) *
+          μ (x ^ (n % m1)) ^ (1 / (n : ℝ)) :=
+        (mul_rpow (apply_nonneg μ _) (apply_nonneg μ _))
+      _ ≤ (μ (x ^ (m1 : ℕ)) ^ (n / (m1 : ℕ))) ^ (1 / (n : ℝ)) *
+            μ (x ^ (n % m1)) ^ (1 / (n : ℝ)) := ((mul_le_mul_right h4).mpr h)
+      _ < (L + ε / 2) * (L + ε / 2) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) *
+            μ (x ^ (n % m1)) ^ (1 / (n : ℝ)) := (mul_lt_mul h1 (le_refl _) h4 (le_of_lt h5))
+      _ ≤ (L + ε / 2) * (L + ε / 2) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) *
+            (μ x ^ (n % m1)) ^ (1 / (n : ℝ)) := ((mul_le_mul_left h5).mpr h2)
+      _ ≤ L + ε := h3
+
+/-- If `μ 1 ≤ 1`, then `smoothingFun μ x` is the limit of `smoothingSeminormSeq μ x`
+  as `n` tends to infinity. -/
+theorem tendsto_smoothingFun_of_map_one_le_one (hμ1 : μ 1 ≤ 1) (x : R) :
+    Tendsto (smoothingSeminormSeq μ x) atTop (𝓝 (smoothingFun μ x)) := by
+  by_cases hx : μ x = 0
+  · exact tendsto_smoothingFun_of_eq_zero μ hx
+  · exact tendsto_smoothingFun_of_ne_zero μ hμ1 hx
+
+/-- If `μ 1 ≤ 1`, then `smoothingFun μ x` is nonnegative. -/
+theorem smoothingFun_nonneg (hμ1 : μ 1 ≤ 1) (x : R) : 0 ≤ smoothingFun μ x := by
+  apply ge_of_tendsto (tendsto_smoothingFun_of_map_one_le_one μ hμ1 x)
+  simpa only [eventually_atTop, ge_iff_le] using ⟨1, fun _ _ ↦ rpow_nonneg (apply_nonneg μ _) _⟩
+
+/-- If `μ 1 ≤ 1`, then `smoothingFun μ 1 ≤ 1`. -/
+theorem smoothingFun_one_le (hμ1 : μ 1 ≤ 1) : smoothingFun μ 1 ≤ 1 := by
+  apply le_of_tendsto (tendsto_smoothingFun_of_map_one_le_one μ hμ1 (1 : R))
+  simp only [eventually_atTop, ge_iff_le]
+  use 1
+  rintro n hn
+  simp only [smoothingSeminormSeq]
+  rw [one_pow]
+  conv_rhs => rw [← one_rpow (1 / n : ℝ)]
+  have hn1 : 0 < (1 / n : ℝ) := by
+    apply div_pos zero_lt_one
+    rw [← cast_zero, cast_lt]
+    exact succ_le_iff.mp hn
+  exact (rpow_le_rpow_iff (apply_nonneg μ _) zero_le_one hn1).mpr hμ1
+
+/-- For any `x` and any positive `n`, `smoothingFun μ x ≤ μ (x ^ (n : ℕ))^(1 / n : ℝ)`. -/
+theorem smoothingFun_le (x : R) (n : PNat) :
+    smoothingFun μ x ≤ μ (x ^ (n : ℕ)) ^ (1 / n : ℝ) :=
+  ciInf_le (smoothingSeminormSeq_bddBelow μ x) _
+
+/-- For all `x : R`, `smoothingFun μ x ≤ μ x`. -/
+theorem smoothingFun_le_self (x : R) : smoothingFun μ x ≤ μ x := by
+  apply (smoothingFun_le μ x 1).trans
+  rw [PNat.one_coe, pow_one, cast_one, div_one, rpow_one]
+
+/- In this section, we prove that if `μ` is nonarchimedean, then `smoothingFun μ` is
+  nonarchimedean. -/
+section IsNonarchimedean
+
+/-- Auxiliary sequence for the proof that `smoothingFun` is nonarchimedean. -/
+private def mu {x y : R} (hn : ∀ n : ℕ, ∃ m < n + 1,
```

Comment:
Why don't you itnroduce `hn` as a `variable` so that you can remove it from the lemmas below?